# Hypothesis Testing: Unveiling Hidden Truths in Your Data

In the world of data analysis, hypothesis testing is a powerful tool used to make informed decisions about entire populations (like all your customers) based on the data we collect from a smaller sample. It works by setting up two competing theories:

- The Null Hypothesis (H0): This is the "business as usual" scenario, assuming there's no significant effect or relationship between the variables we're studying.
- The Alternative Hypothesis (H1): This is what we're actually interested in – the theory we want to prove or disprove using the data.

By analyzing sample data, we can assess the likelihood of observing the results we have if the null hypothesis were true. If this probability falls below a certain threshold (typically 5%), we can reject the null hypothesis and conclude that there's likely a real effect or relationship at play, supporting our alternative hypothesis.

**Focus Here:** This approach allows us to move beyond assumptions and base our decisions on statistically significant evidence within the data itself.

Hypothesis testing empowers data analysts to make data-driven decisions about entire populations. By analyzing a smaller, representative sample, we can infer trends and relationships that hold true for the larger group.

This structured approach follows a clear path:

1. Gather Data: We collect relevant data for the hypothesis.
2. Formulate Hypotheses: We define opposing null (H0) and alternative (H1) hypotheses.
3. Set Significance Level: We establish the acceptable risk of rejecting a true null hypothesis (α).
4. Choose the Right Test: We select the appropriate statistical test based on data type and research question.
5. Run the Test: We apply the chosen test to our data.
6. Interpret the P-Value: The p-value tells us the likelihood of observing the data if the null hypothesis were true. A low p-value suggests rejecting H0 and supporting H1.
Equipped with a well-defined plan and the right data, we're ready to delve into the analysis and unlock the hidden truths within our sample.

To get started with Hypothesis Testing, we need appropriate data. I found an ideal dataset for this task. You can download the dataset from [here](https://statso.io/light-theme-and-dark-theme-case-study/).

#### Hypothesis Testing using Python
Now, let’s get started with the task of Hypothesis Testing by importing the necessary Python libraries and the dataset:

In [1]:
import pandas as pd
from scipy.stats import ttest_ind

df = pd.read_csv("website_ab_test.csv")

print(df.head())

         Theme  Click Through Rate  Conversion Rate  Bounce Rate  \
0  Light Theme            0.054920         0.282367     0.405085   
1  Light Theme            0.113932         0.032973     0.732759   
2   Dark Theme            0.323352         0.178763     0.296543   
3  Light Theme            0.485836         0.325225     0.245001   
4  Light Theme            0.034783         0.196766     0.765100   

   Scroll_Depth  Age   Location  Session_Duration Purchases Added_to_Cart  
0     72.489458   25    Chennai              1535        No           Yes  
1     61.858568   19       Pune               303        No           Yes  
2     45.737376   47    Chennai               563       Yes           Yes  
3     76.305298   58       Pune               385       Yes            No  
4     48.927407   25  New Delhi              1437        No            No  


Our website aims to optimize user experience, and a key factor is theme design. To determine which of two themes (Theme A vs. Theme B) performs better, we'll leverage hypothesis testing.

This initial step involves examining the dataset we have at hand. Let's delve into the data summary:

Record Count, Missing Values, and Numerical Statistics of the colunms
By analyzing these data characteristics, we can establish a solid foundation for formulating our hypotheses and selecting the appropriate statistical test for our website theme performance evaluation.

In [2]:
# dataset summary
summary = {
    'Number of Records': df.shape[0],
    'Number of Columns': df.shape[1],
    'Missing Values': df.isnull().sum(),
    'Numerical Columns Summary': df.describe()
}

summary

{'Number of Records': 1000,
 'Number of Columns': 10,
 'Missing Values': Theme                 0
 Click Through Rate    0
 Conversion Rate       0
 Bounce Rate           0
 Scroll_Depth          0
 Age                   0
 Location              0
 Session_Duration      0
 Purchases             0
 Added_to_Cart         0
 dtype: int64,
 'Numerical Columns Summary':        Click Through Rate  Conversion Rate  Bounce Rate  Scroll_Depth  \
 count         1000.000000      1000.000000  1000.000000   1000.000000   
 mean             0.256048         0.253312     0.505758     50.319494   
 std              0.139265         0.139092     0.172195     16.895269   
 min              0.010767         0.010881     0.200720     20.011738   
 25%              0.140794         0.131564     0.353609     35.655167   
 50%              0.253715         0.252823     0.514049     51.130712   
 75%              0.370674         0.373040     0.648557     64.666258   
 max              0.499989         0.49891

The dataset contains 1,000 records across 10 columns, with no missing values. Here’s a quick summary of the numerical columns:

- **Click Through Rate:** Ranges from about 0.01 to 0.50 with a mean of approximately 0.26.
- **Conversion Rate:** Also ranges from about 0.01 to 0.50 with a mean close to the Click Through Rate, approximately 0.25.
- **Bounce Rate:** Varies between 0.20 and 0.80, with a mean around 0.51.
- **Scroll Depth:** Shows a spread from 20.01 to nearly 80, with a mean of 50.32.
- **Age:** The age of users ranges from 18 to 65 years, with a mean age of about 41.5 years.
- **Session Duration:** This varies widely from 38 seconds to nearly 1800 seconds (30 minutes), with a mean session duration of approximately 925 seconds (about 15 minutes). 

Now, Let's head to comparing both themes perfromance using the provided metrics. We shall exmine the average Click Through Rate, Conversion Rate, Bounce Rate, and other relevant metrics for each theme. Following that, we can perform hypothesis testing to identify if there’s a statistically significant difference between the themes 

In [3]:
# grouping data by theme and calculating mean values for the metrics
theme_performance = df.groupby('Theme').mean()

# sorting the data by conversion rate for a better comparison
theme_performance_sorted = theme_performance.sort_values(by='Conversion Rate', ascending=False)

print(theme_performance_sorted)

             Click Through Rate  Conversion Rate  Bounce Rate  Scroll_Depth  \
Theme                                                                         
Light Theme            0.247109         0.255459     0.499035     50.735232   
Dark Theme             0.264501         0.251282     0.512115     49.926404   

                   Age  Session_Duration  
Theme                                     
Light Theme  41.734568        930.833333  
Dark Theme   41.332685        919.482490  


C:\Users\Sroor For Laptop\AppData\Local\Temp\ipykernel_21020\2531333806.py:2: FutureWarning: The default value of numeric_only in DataFrameGroupBy.mean is deprecated. In a future version, numeric_only will default to False. Either specify numeric_only or select only columns which should be valid for the function.
  theme_performance = df.groupby('Theme').mean()


The comparison between the Light Theme and Dark Theme on average performance metrics reveals the following insights:

- **Click Through Rate (CTR):** The Dark Theme has a slightly higher average CTR (0.2645) compared to the Light Theme (0.2471).
- **Conversion Rate:** The Light Theme leads with a marginally higher average Conversion Rate (0.2555) compared to the Dark Theme (0.2513).
- **Bounce Rate:** The Bounce Rate is slightly higher for the Dark Theme (0.5121) than for the Light Theme (0.4990).
- **Scroll Depth:** Users on the Light Theme scroll slightly further on average (50.74%) compared to those on the Dark Theme (49.93%).
- **Age:** The average age of users is similar across themes, with the Light Theme at approximately 41.73 years and the Dark Theme at 41.33 years.
- **Session Duration:** The average session duration is slightly longer for users on the Light Theme (930.83 seconds) than for those on the Dark Theme (919.48 seconds).

Based on these insight, The Light Theme appears to slightly outproforms the Dark Theme when it comes to Conversion Rate, Bounce Rate, Scroll Depth, and Session Duration. On the other hand, the Dark Theme leads in Click Through Rate. It is clear that the difference are relatively minor across all metrics.

#### Getting Started with Hypothesis Testing
A significance level (alpha) of 0.05 will be used for the hypothesis testing. This means if the p-value is less than 0.05, The result is considered statistically significant.

Let’s start with hypothesis testing based on the Conversion Rate between the Light Theme and Dark Theme. Our hypotheses are as follows:

- Null Hypothesis (H0​): There is no difference in Conversion Rates between the Light Theme and Dark Theme.
- Alternative Hypothesis (Ha​): There is a difference in Conversion Rates between the Light Theme and Dark Theme.

We’ll use a two-sample t-test to compare the means of the two independent samples. Let’s proceed with the test:

In [4]:
# extracting conversion rates for both themes
conversion_rates_light = df[df['Theme'] == 'Light Theme']['Conversion Rate']
conversion_rates_dark = df[df['Theme'] == 'Dark Theme']['Conversion Rate']

# performing a two-sample t-test
t_stat, p_value = ttest_ind(conversion_rates_light, conversion_rates_dark, equal_var=False)

t_stat, p_value

(0.4748494462782632, 0.6349982678451778)

The two-sample t-test results in p-value of 0.635 which is much greater than the significance level of 0.05. We do not have enough evidence to reject the null hypothesis. Hence, we conclude that there is no statistically significant difference in Conversion Rates between the Light Theme and Dark Theme based on the data provided.

Now, let’s move to conduct hypothesis testing based on the Click Through Rate (CTR) to see wheather there’s a statistically significant difference between the Light Theme and Dark Theme regarding how often users click through. Our hypotheses remain structured similarly:

- Null Hypothesis (H0​): There is no difference in Click Through Rates between the Light Theme and Dark Theme.
- Alternative Hypothesis (Ha​): There is a difference in Click Rates between the Light Theme and Dark Theme.

We’ll perform a two-sample t-test on the CTR for both themes. Let’s proceed with the calculation:

In [5]:
# extracting click through rates for both themes
ctr_light = df[df['Theme'] == 'Light Theme']['Click Through Rate']
ctr_dark = df[df['Theme'] == 'Dark Theme']['Click Through Rate']

# performing a two-sample t-test
t_stat_ctr, p_value_ctr = ttest_ind(ctr_light, ctr_dark, equal_var=False)

t_stat_ctr, p_value_ctr

(-1.9781708664172253, 0.04818435371010704)

The two-sample t-test for the Click Through Rate (CTR)results in p-value of 0.048 which is slightly less than the significance level of 0.05. Therefore, there is a statistically significant difference in Click Through Rates between the Light Theme and Dark Theme, with the Dark Theme likely having a higher CTR given the direction of the test statistic.

Now, let’s perform Hypothesis Testing based on two other metrics: 
- bounce rate 
- scroll depth

These two metrics are important metrics for analyzing the performance of a theme or a design on a website. 
Firstly we perform these statistical tests and then create a table to show the report of all the tests we have done:

In [6]:
# extracting bounce rates for both themes
bounce_rates_light = df[df['Theme'] == 'Light Theme']['Bounce Rate']
bounce_rates_dark = df[df['Theme'] == 'Dark Theme']['Bounce Rate']

# performing a two-sample t-test for bounce rate
t_stat_bounce, p_value_bounce = ttest_ind(bounce_rates_light, bounce_rates_dark, equal_var=False)

# extracting scroll depths for both themes
scroll_depth_light = df[df['Theme'] == 'Light Theme']['Scroll_Depth']
scroll_depth_dark = df[df['Theme'] == 'Dark Theme']['Scroll_Depth']

# performing a two-sample t-test for scroll depth
t_stat_scroll, p_value_scroll = ttest_ind(scroll_depth_light, scroll_depth_dark, equal_var=False)

# creating a table for comparison
comparison_table = pd.DataFrame({
    'Metric': ['Click Through Rate', 'Conversion Rate', 'Bounce Rate', 'Scroll Depth'],
    'T-Statistic': [t_stat_ctr, t_stat, t_stat_bounce, t_stat_scroll],
    'P-Value': [p_value_ctr, p_value, p_value_bounce, p_value_scroll]
})

comparison_table

Metric  T-Statistic   P-Value
0  Click Through Rate    -1.978171  0.048184
1     Conversion Rate     0.474849  0.634998
2         Bounce Rate    -1.201888  0.229692
3        Scroll Depth     0.756228  0.449692

As showing on the table comparing the performance of the Light Theme and Dark Theme across various metrics based on hypothesis testing:
- **Click Through Rate:** The test reveals a statistically significant difference, with the Dark Theme likely performing better (P-Value = 0.048).
- **Conversion Rate:** No statistically significant difference was found (P-Value = 0.635).
- **Bounce Rate:** No statistically significant difference in Bounce Rates between the themes (P-Value = 0.230).
- **Scroll Depth:** No statistically significant difference is observed in Scroll Depths (P-Value = 0.450).

### Summary
This concludes that although the two themes show similar performance based on most metrics, the Dark Theme slightly outperforms the Light Thmese n terms of engaging users to click through. the Light Theme and a Dark Theme choice does not significantly affect user behaviour according to the data provided when it comes to other mertics such as Conversion Rate, Bounce Rate, and Scroll Depth
